### PR TITLE
read resource attributes from annotations

### PR DESCRIPTION
**Pull request by zeitlinger**
Fixes https://github.com/open-telemetry/opentelemetry-operator/issues/3112

**Testing:**  Added integration test

**Documentation:** will be added to https://opentelemetry.io/docs/kubernetes/operator/automatic/ once released

### CODE REVIEW

**Comment by jaronoff97** [comment]
per our discussion, @zeitlinger will rebase and expand on the capabilities introduced by #2330 :)

File: pkg/instrumentation/sdk_test.go
Comment (jaronoff97) [inline pkg/instrumentation/sdk_test.go:162]: nit: can we remove the space here

**Review by jaronoff97 — COMMENTED** [review]
Can you also update the README.md, explaining the changes and also explaining the precedence of deciding these variables?

File: pkg/instrumentation/sdk_test.go
Comment (zeitlinger) [reply pkg/instrumentation/sdk_test.go]: done

**Review by jaronoff97 — APPROVED** [review]
@zeitlinger last thing – can you update the README.md with some more information about the new behavior? Maybe also include it in the subtext of the chlog entry

File: pkg/instrumentation/sdk.go
Comment (TylerHelmuth) [inline pkg/instrumentation/sdk.go:425]: The parameter name is `labelKey`, but when invoked the `constants.AnnotationAppName` value is used.  Then when indexing the annotations the resource is used, not the `labelKey`.  Is this wired up correctly?

File: pkg/instrumentation/sdk.go
Comment (TylerHelmuth) [inline pkg/instrumentation/sdk.go:400]: I am concerned that this would be a breaking change.

File: pkg/instrumentation/sdk.go
Comment (TylerHelmuth) [inline pkg/instrumentation/sdk.go:456]: I'm concerned that this would be a breaking change.

File: pkg/instrumentation/sdk.go
Comment (zeitlinger) [reply pkg/instrumentation/sdk.go:400]: yes, I'll include that in the changelog

File: pkg/instrumentation/sdk.go
Comment (zeitlinger) [reply pkg/instrumentation/sdk.go:425]: labelkey is looked up in labels:

https://github.com/open-telemetry/opentelemetry-operator/blob/bef08935239ce6c7859fa9c0d7ee59f074c8c306/pkg/instrumentation/sdk.go#L429

**Comment by zeitlinger** [comment]
> @zeitlinger last thing – can you update the README.md with some more information about the new behavior? Maybe also include it in the subtext of the chlog entry

done

**Review by pavolloffay — APPROVED** [review]

File: pkg/instrumentation/sdk.go
Comment (TylerHelmuth) [reply pkg/instrumentation/sdk.go:400]: @jaronoff97 @pavolloffay I am pretty worried about this breaking change.  Changes to `service.name` will change emitted telemetry, which could affect end-users alerts/SLOs/dashboards/etc. This change will (likely) not  be detected with unit or e2e tests before a user updates their operator. Additionally, the operator's auto-instrumentation injection is already a huge obfuscation on top of already obfuscated auto-instrumentation, which means that our end-users are even less likely to understand the impact of this change.

I would much rather prefer this work be implemented in such a way that the annotation is used as the service name if no other service name is set OR if some other explicit configuration field is set that enables this behavior.

Comment (TylerHelmuth) [reply pkg/instrumentation/sdk.go:400]: The reason I call this out as a breaking change in the current implementation is that it is now defaulting to the annotation value if it exists.  Since it is a common annotation, it will probably be on most end-user's pods, resulting in the potential breaking change.

I would much rather this be the last option in this list to avoid the breaking change.

Comment (TylerHelmuth) [reply pkg/instrumentation/sdk.go:400]: In addition, there is no stable semantic convention that ranks which value should be used as the `service.name`. I'd prefer avoiding breaking changes on this topic until there is a stable semantic convention, so that we can minimize the number of breaking changes.

Comment (zeitlinger) [reply pkg/instrumentation/sdk.go:400]: we could solve this problem if we let the user define the mapping they prefer - so it would be an opt-in feature

Comment (jaronoff97) [reply pkg/instrumentation/sdk.go:400]: from SIG:
I think we should move forward with this by putting it within a new "defaults" section in the instrumentation CRD as so:
```
apiVersion: opentelemetry.io/v1alpha1
kind: Instrumentation
metadata:
  name: my-instrumentation
spec:
  defaults:
    useLabelsForResource: true
```

We also must then document in the readme the precedence for the automatic variables we set for users and how this boolean changes that.

File: pkg/instrumentation/sdk.go
Comment (zeitlinger) [reply pkg/instrumentation/sdk.go]: changed constants to have `label` prefix

File: pkg/instrumentation/sdk.go
Comment (zeitlinger) [reply pkg/instrumentation/sdk.go]: fixed

**Comment by pavolloffay** [comment]
CI failed

File: README.md
Comment (jaronoff97) [inline README.md:739]: let's have this under a new section

File: README.md
Comment (jaronoff97) [inline README.md:783]: let's change this to be a numbered list

Comment (jaronoff97) [reply README.md:783]: where does the new `useLabelsForResourceAttributes` fall into this? it is the last rule? if so, we should specify that you need to set defaults.useLabelsForResourceAttributes=true

File: apis/v1alpha1/instrumentation_types.go
Comment (jaronoff97) [inline apis/v1alpha1/instrumentation_types.go:122]: let's include the mapping rules in this doc string too

File: pkg/instrumentation/apachehttpd.go
Comment (jaronoff97) [inline pkg/instrumentation/apachehttpd.go:98]:
```suggestion
		// and we definitely do not need them for the init container for cp command
```

File: pkg/instrumentation/sdk.go
Comment (jaronoff97) [inline pkg/instrumentation/sdk.go:297]: we shouldn't need the bool flag on this method if it's already in the otelinst object

File: pkg/instrumentation/sdk.go
Comment (jaronoff97) [inline pkg/instrumentation/sdk.go:416]: can we add a docstring for this method that includes the precedence?

Comment (zeitlinger) [reply pkg/instrumentation/sdk.go:416]: also added a docstring to similar methods

**Comment by zeitlinger** [comment]
@jaronoff97 please check again

File: README.md
Comment (jaronoff97) [inline README.md:720]:
```suggestion
## Configure resource attributes
```

File: pkg/instrumentation/sdk.go
Comment (jaronoff97) [inline pkg/instrumentation/sdk.go:499]: as far as precedence goes, I do think we want this to remain last right? i.e. if someone is explicitly setting `"resource.opentelemetry.io/service.name": "my-service"` I think that should have final say. Otherwise, this would be a breaking change

**Review by jaronoff97 — COMMENTED** [review]
another thought, also if we could add something in to an existing e2e test that would be great :) thank you! FWIW, i do think I will be taking on a refactor to make this logic a bit more straightforward in the future.

File: pkg/instrumentation/sdk.go
Comment (zeitlinger) [reply pkg/instrumentation/sdk.go:499]: right - also for the labels

Comment (zeitlinger) [reply pkg/instrumentation/sdk.go:499]: done

**Comment by zeitlinger** [comment]
> another thought, also if we could add something in to an existing e2e test that would be great :) 

done @jaronoff97

File: pkg/instrumentation/sdk.go
Comment (jaronoff97) [inline pkg/instrumentation/sdk.go:548]: why not use the method call here?

File: pkg/instrumentation/sdk.go
Comment (jaronoff97) [inline pkg/instrumentation/sdk.go:550]: we shouldn't modify the iteration variable here

File: pkg/instrumentation/sdk.go
Comment (jaronoff97) [inline pkg/instrumentation/sdk.go:551]: don't we want to let the override happen here?

File: pkg/instrumentation/sdk.go
Comment (zeitlinger) [reply pkg/instrumentation/sdk.go:548]: which method call?

File: pkg/instrumentation/sdk.go
Comment (zeitlinger) [reply pkg/instrumentation/sdk.go:551]: we treat the user defined enviroment variables as the highest priority - otherwise it would be a breaking change.
Not sure if it would also be surprising.

https://github.com/zeitlinger/opentelemetry-operator/blob/c1966659f203516c87991715c95f1dc00a372aa1/README.md?plain=1#L783

**Comment by zeitlinger** [comment]
@jaronoff97 please check again

File: pkg/instrumentation/sdk.go
Comment (jaronoff97) [reply pkg/instrumentation/sdk.go:548]: `pod.GetAnnotations()`

File: pkg/instrumentation/sdk_test.go
Comment (jaronoff97) [inline pkg/instrumentation/sdk_test.go:532]: OOC – why change this for the test?

**Review by jaronoff97 — APPROVED** [review]
two minor questions, neither blocking. overall i think this looks good. Going to wait on @TylerHelmuth's review before merging.

File: pkg/instrumentation/sdk_test.go
Comment (zeitlinger) [reply pkg/instrumentation/sdk_test.go:532]: to make sure it's coming from the right place

File: tests/e2e/env-vars/00-install-instrumentation.yaml
Comment (TylerHelmuth) [inline tests/e2e/env-vars/00-install-instrumentation.yaml:9]: Please add a new test with this flag set, I'd like to preserve the test that uses the existing behavior

Comment (zeitlinger) [reply tests/e2e/env-vars/00-install-instrumentation.yaml]: done @TylerHelmuth

**Review by TylerHelmuth — APPROVED** [review]

**Review by swiatekm — APPROVED** [review]

**Comment by jaronoff97** [comment]
@zeitlinger thank you for your contribution! 🙇

**Pull request merged**
